### PR TITLE
[Bugfix/Enhancement] Find files with a beginning character match (in Windows dialog)

### PR DESCRIPTION
**Pull request by anthonyleestark**
## 🐞Description of the Issue
In the Windows dialog, pressing a character key jumps to the first item instead of to files beginning with that character, unlike other applications.

## 🔍 Cause and Solution
1. The Windows dialog uses the standard `SysListView32` control. 
2. By default, the `SysListView32` control handles the `WM_CHAR` message by searching for the first item that starts with the typed character and jumping to it.
3. In the Windows dialog, the list view has focus, and the `WM_CHAR` message is sent directly to it, but it does not behave as expected. For unknown reasons, it always jumps to the first item, regardless of the actual character typed.
4. So to work around this issue, I installed a message hook to intercept the `WM_CHAR` message **before** it reaches the list view. Instead of letting the control handle it, I forward a replacement of the `WM_CHAR` message directly to the dialog and implement custom handling there.

## ✨ What's New in This Enhancement
1. **Fix the issue where pressing a character key always jumps to the first item**
2. **Implement character key-based item search with the following behavior:**
2.1. By default, pressing a character key jumps to the first item that begins with that character.
2.2. If there are multiple matching items and one or more of them are already selected, pressing the key again will jump to the next matching item after the **last selected** one.
2.3. If all matching items are already selected, pressing the key will loop back and jump to the **first** matching item.
2.4. If **no matching item** is found, no jump occurs and **a beep sound** is triggered to indicate no match.

## ✅ Test Plan
- Manual testing on Windows 10/11
- Verified edge cases with no matching items
- Tested behavior with all items selected

## 🔗 Related Issue
Merging this PR will close the issue #2239 

## 📋 Checklist
- [x] Code compiles without errors
- [x] Manual testing completed
- [x] No regressions observed

### CODE REVIEW

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 1]
```diff
@@ -484,6 +516,99 @@ intptr_t CALLBACK WindowsDlg::run_dlgProc(UINT message, WPARAM wParam, LPARAM lP
...
+
+					// case 3:
+					else
+					{
+						// restore previous state
+						for (int index = 0; index < itemCount; index++)
+							ListView_SetItemState(_hList, index, stateMap.at(index), LVIS_SELECTED | LVIS_FOCUSED);
+
+						// make a beep sound
+						MessageBeep(0xFFFFFFFF);
```
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:604]: Use `_muteSounds` (please search in the project to know the usage).

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 2]
```diff
@@ -484,6 +516,99 @@ intptr_t CALLBACK WindowsDlg::run_dlgProc(UINT message, WPARAM wParam, LPARAM lP
 			break;
 		}
 
+		case WM_CHAR_REPLACEMENT:
+		{
+			if (_hList != NULL)
+			{
```
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:522]: ...to remove '{' for reducing the depth of indentation, so the code is more readable.

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 3]
```diff
@@ -484,6 +516,99 @@ intptr_t CALLBACK WindowsDlg::run_dlgProc(UINT message, WPARAM wParam, LPARAM lP
 			break;
 		}
 
+		case WM_CHAR_REPLACEMENT:
+		{
+			if (_hList != NULL)
+			{
+				wchar_t ch = (wchar_t)wParam;
```
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:523]: Use C++ conversion convention.

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 4]
```diff
@@ -484,6 +516,99 @@ intptr_t CALLBACK WindowsDlg::run_dlgProc(UINT message, WPARAM wParam, LPARAM lP
...
+				// backup current state
+				vector<int> stateMap;
+				stateMap.reserve(itemCount);
+				for (int index = 0; index < itemCount; index++)
+					stateMap.push_back(ListView_GetItemState(_hList, index, LVIS_SELECTED | LVIS_FOCUSED));
+
+				// temporarily clear existing selection
+				ListView_SetItemState(_hList, -1, 0, LVIS_SELECTED | LVIS_FOCUSED);
+
+				LVITEM lvItem = { 0 };
```
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:535]: Respect the coding style in the project: Use `LVITEM lvItem {};` instead.

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 5]
```diff
@@ -484,6 +516,99 @@ intptr_t CALLBACK WindowsDlg::run_dlgProc(UINT message, WPARAM wParam, LPARAM lP
...
+				wchar_t buffer[MAX_PATH];
```
Comment (donho) [inline PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:536]: No initialization.

**Review by donho — CHANGES_REQUESTED** [review]
Please follow the suggestions:

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp [hunk 6]
```diff
@@ -42,6 +42,9 @@ using namespace std;
 #define WD_MENUCOPYNAME				"MenuCopyName"
 #define WD_MENUCOPYPATH				"MenuCopyPath"
 
+// WM_CHAR replacement
+#define WM_CHAR_REPLACEMENT			(WM_USER+WM_CHAR)
```
[the remark this answers is not in the document]
Comment (anthonyleestark) [reply PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp:46]: I'm using a replacement for `WM_CHAR` to prevent an infinite message loop (in the `getMsgProc` function).
Since this replacement is only used within the Windows dialog, it could be defined as `WM_USER + <any arbitrary number>`. 
I chose `WM_USER + WM_CHAR` to make it clear that **this message is a replacement for `WM_CHAR`**

**Comment by anthonyleestark** [comment]
@donho 
I've already updated the review changes

File: PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp (hunk 3 above)
Comment (donho) [reply PowerEditor/src/WinControls/WindowsDlg/WindowsDlg.cpp]: By "Use C++ conversion convention." I mean:
Use `wchar_t ch = static_cast<wchar_t>(wParam);` instead.

**Review by donho — CHANGES_REQUESTED** [review]
Use `wchar_t ch = static_cast<wchar_t>(wParam);` instead of `wchar_t ch = (wchar_t)wParam;`.

**Comment by donho** [comment]
This should not happen:
[Merge remote-tracking branch 'origin/master' into enhancement/issue-2239](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/16572/commits/7a9282ab493649f9fce2d5cf75c25c23e3f15790)

***Don't merge upstream/master (using git or via github sync), if your PR is based on older upstream/master. If you need to base it on latest master (e.g. to check and fix merge conflict), use commands git fetch upstream to get latest master and then git rebase upstream/master to rebase it onto this latest upstream/master.***
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#guidelines-for-pull-requests

**Comment by anthonyleestark** [comment]
@donho 
I've already done that, please check it again

**Comment by anthonyleestark** [comment]
> This should not happen: [Merge remote-tracking branch 'origin/master' into enhancement/issue-2239](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/16572/commits/7a9282ab493649f9fce2d5cf75c25c23e3f15790)
> 
> _**Don't merge upstream/master (using git or via github sync), if your PR is based on older upstream/master. If you need to base it on latest master (e.g. to check and fix merge conflict), use commands git fetch upstream to get latest master and then git rebase upstream/master to rebase it onto this latest upstream/master.**_ https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#guidelines-for-pull-requests

My mistake, sorry
